core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/PushJoinIntoTableScan.java

```diff
@@ -123,31 +147,54 @@ public Result apply(JoinNode joinNode, Captures captures, Context context)
 
         TableHandle handle = joinApplicationResult.get().getTableHandle();
 
-        Map<ColumnHandle, ColumnHandle> columnHandlesMapping = joinApplicationResult.get().getColumnHandles();
-
-        TupleDomain<ColumnHandle> newEnforcedConstraints = TupleDomain.fromColumnDomains(
-                transformToNewAssignments(left.getEnforcedConstraint(), columnHandlesMapping).getColumnDomains().flatMap(
-                        leftDomains -> transformToNewAssignments(right.getEnforcedConstraint(), columnHandlesMapping).getColumnDomains().flatMap(
-                                rightDomains -> Optional.of(ImmutableList.<TupleDomain.ColumnDomain<ColumnHandle>>builder().addAll(leftDomains).addAll(rightDomains).build()))));
+        Map<ColumnHandle, ColumnHandle> leftColumnHandlesMapping = joinApplicationResult.get().getLeftColumnHandles();
+        Map<ColumnHandle, ColumnHandle> rightColumnHandlesMapping = joinApplicationResult.get().getRightColumnHandles();
 
         ImmutableMap.Builder<Symbol, ColumnHandle> newAssignments = ImmutableMap.builder();
         newAssignments.putAll(left.getAssignments().entrySet().stream().collect(toImmutableMap(
                 Map.Entry::getKey,
-                entry -> columnHandlesMapping.get(entry.getValue()))));
+                entry -> leftColumnHandlesMapping.get(entry.getValue()))));
         newAssignments.putAll(right.getAssignments().entrySet().stream().collect(toImmutableMap(
                 Map.Entry::getKey,
-                entry -> columnHandlesMapping.get(entry.getValue()))));
+                entry -> rightColumnHandlesMapping.get(entry.getValue()))));
+
+        // convert enforced constraint
+        TupleDomain<ColumnHandle> leftConstraint = left.getEnforcedConstraint();
+        TupleDomain<ColumnHandle> rightConstraint = right.getEnforcedConstraint();
+        if (joinNode.getType() == LEFT || joinNode.getType() == FULL) {
+            leftConstraint = leftConstraint.transformDomains((columnHandle, domain) -> domain.union(onlyNull(domain.getType())));
+        }
+        if (joinNode.getType() == RIGHT || joinNode.getType() == FULL) {
+            rightConstraint = rightConstraint.transformDomains((columnHandle, domain) -> domain.union(onlyNull(domain.getType())));
+        }
+        leftConstraint = transformToNewAssignments(leftConstraint, leftColumnHandlesMapping);
```

Comment:
The code would be clearer if the logic was extracted:

```java
JoinNode.Type joinType = joinNode.getType();
TupleDomain<ColumnHandle> leftConstraint = deriveConstraint(left.getEnforcedConstraint(), leftColumnHandlesMapping, joinType == RIGHT || joinType == FULL);
TupleDomain<ColumnHandle> rightConstraint = deriveConstraint(right.getEnforcedConstraint(), rightColumnHandlesMapping, joinType == LEFT || joinType == FULL);

TupleDomain<ColumnHandle> newEnforcedConstraint = TupleDomain.withColumnDomains(
        ImmutableMap.<ColumnHandle, Domain>builder()
                // we are sure that domains map is present as we bailed out on isNone above
                .putAll(leftConstraint.getDomains().orElseThrow())
                .putAll(rightConstraint.getDomains().orElseThrow())
                .build());
```

```java
private TupleDomain<ColumnHandle> deriveConstraint(TupleDomain<ColumnHandle> sourceConstraint, Map<ColumnHandle, ColumnHandle> columnMapping, boolean nullable)
{
    TupleDomain<ColumnHandle> constraint = sourceConstraint;
    if (nullable) {
        constraint = constraint.transformDomains((columnHandle, domain) -> domain.union(onlyNull(domain.getType())));
    }
    return constraint.transform(handle -> {
        ColumnHandle newHandle = columnMapping.get(handle);
        checkArgument(newHandle != null, "Mapping not found for handle %s", handle);
        return newHandle;
    });
}
```

Reply:
cleaner indeed. THanks.